src/blog/system-analyzer.md

```diff
@@ -0,0 +1,206 @@
+
+![System Analyzer Overview](/_img/system-analyzer/system-analyzer-overview.png)
+
+![Group IC events by function name to get in depth information about the IC events associated with the `dotProduct`.](/_img/system-analyzer/case1_1.png)
+
+We are analyzing how the function `dotProduct` might be causing this performance difference. So we group IC events by functionName to get more in depth information about IC events associated with the `dotProduct` function.
+
+The first thing we notice is that we have two different IC state transitions recorded by the IC events in this function. One going from uninitialised to monomorphic and the other one going from monomorphic to polymorphic. As mentioned in the previously linked [blog post](https://mathiasbynens.be/notes/shapes-ics), polymorphic IC state indicates that now we are tracking more than one Map associated with `Point` objects.
+
+We want to know why we are creating multiple Map shapes for the same type of objects we toggle the info button about IC state to get more information about the Map addresses going from uninitialised to monomorphic.
+
+![The map transition tree associated with the monomorphic IC state.](/_img/system-analyzer/case1_2.png)
+
+![The map transition tree associated with the polymorphic IC state.](/_img/system-analyzer/case1_3.png)
+
+For the monomorphic IC state we can visualise the transition tree and see that we are only dynamically adding two properties `x`and `y` but when it comes to polymorphic IC state, we have a new Map containing three properties `isNegative`, `x` and `y`.
+
+![The Map panel communicates the file position information to highlight file positions on the Source panel.](/_img/system-analyzer/case1_4.png)
+
+We click on the file position section of the Map panel to see where this `isNegative` property is added in the source code.
+
+So now the question being *how can we address the performance regression by using the insight we generated from the tool*?
+
+The minimal solution would be to always initialise the `isNegative` property. As a common advice we say that all instance properties should be initialised in the constructor.
+
+Now, the updated `Point` class looks like this:
+
+```javascript
+class Point {
+ constructor(x, y) {
+   this.isNegative = x < 0 || y < 0;
+   this.x = x;
+   this.y = y;
+ }
+
+ dotProduct(other) {
+   return this.x * other.x + this.y * other.y;
+ }
+}
+```
+
+If we execute the script again with the modified `Point` class, we see on the that execution of the two snippets we defined at the beginning of the case study perform very similarly.
+
+![Performance analysis of snippets.](/_img/system-analyzer/final-program-performance.png)
+
+![The map transition tree of the modified Point object.](/_img/system-analyzer/case2_1.png)
+
+Then we group IC events by `dotProduct` key in the IC panel. And now, we see that the polymorphic IC state is avoided as we are not creating multiple maps for the same type of objects.
```

Comment:
Move this above the image

Reply:
In fact, I'd just keep "In an updated trace, we see that the polymorphic IC state is avoided as we are not creating multiple maps for the same type of objects." at the end of the paragraph above.

Reply:
Done